core/src/main/java/ai/timefold/solver/core/api/score/stream/bi/BiConstraintStream.java

```diff
@@ -1582,9 +1583,21 @@ <C> TriConstraintStream<A, B, C> concat(@NonNull TriConstraintStream<A, B, C> ot
             @NonNull Function<A, B> paddingFunction) {
         var firstStream = this;
         var remapped = firstStream.map(ConstantLambdaUtils.biPickFirst());
-        var secondStream = getConstraintFactory().forEach(otherClass)
-                .ifNotExists(remapped, Joiners.equal());
-        return firstStream.concat(secondStream, paddingFunction);
+
+        if (firstStream instanceof AbstractConstraintStream<?> abstractConstraintStream) {
+            var secondStream = switch (abstractConstraintStream.getRetrievalSemantics()) {
+                case STANDARD, LEGACY -> getConstraintFactory().forEach(otherClass);
```

Comment:
The only way how the code could get here through `LEGACY` is if the user somehow used `from` instead of `forEach` -  this has been deprecated for so long that I'd rather throw an exception here. Also, it may even be impossible state, because in this particular situation (`complement`) it is us who specify what operation will be used.

Reply:
Not impossible state, and would be breaking existing behaviour; i.e.

```java
constraintFactory.from(Shift.class)
    .groupBy(Shift::getEmployee, count())
    .complement(Employee.class, employee -> 0)
```
works currently, and would trigger the proposed exception.